File: files/400-499/459/CH-459.R
library(tidyverse)
library(readxl)

path <- "400-499/459/CH-459 Filter.xlsx"
input <- read_excel(path, range = "B3:D13")
test <- read_excel(path, range = "H3:I7")

daily <- input %>%
  summarise(Sales = sum(Sales), .by = c(Product, Date))

qualified <- daily %>%
  group_by(Product) %>%
  summarise(
    qualified = any(map_lgl(seq_len(n()), \(i) {
      i + 2 <= n() && all(diff(Date[i:(i + 2)]) == 1)
    })),
    .groups = "drop"
  ) %>%
  filter(qualified)

result <- daily %>%
  semi_join(qualified, by = "Product") %>%
  arrange(Date) %>%
  select(Date, Product)
all.equal(result, test)

File: files/400-499/459/CH-459.py
import pandas as pd

path = "400-499/459/CH-459 Filter.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=2, nrows=10)
test = pd.read_excel(path, usecols="H:I", skiprows=2, nrows=4)
test.columns = ["Date", "Product"]

daily = input.groupby(["Product", "Date"], as_index=False).Sales.sum()
qualified = daily.groupby("Product").Date.apply(
    lambda d: any(
        pd.Series(sorted(d.unique()))
        .diff()
        .dt.days.eq(1)
        .rolling(2)
        .sum()
        .eq(2)
        .fillna(False)
    )
)
qualified = qualified[qualified].index
result = (
    daily.loc[daily.Product.isin(qualified), ["Date", "Product"]]
    .sort_values("Date")
    .reset_index(drop=True)
)
print(result.equals(test))
